Added grids for Iceland

dalia2 opened the pull request:
added the grids to europe directory, editted README.EUROPE and expected_europe.lst

Commit:
added gris for Iceland

added the grids to europe directory, eddited README.EUROPE and expected_europe.lst

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+### Iceland: Deformation model: ISN93 -> ISN2016 
+
+*Source*: [IGN France](https://geodesie.ign.fr/index.php?page=grilles)   
```
I doubt IGN France authors Icelandic grids ;-)

dalia2 replied:
I have edited the file.

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+The transformation is designed so that the ISN2016 points should fit within 1-2 cm accuracy, but in the area around the earthquake area in the South and North of Vatnajökull the accuracy is worse.
+
+Recommended to transform from ISN93 2D geodetic (EPSG:4659) to ISN2016 2D geodetic (EPSG:8086), also this transformation has code in EPSG Dataset (EPSG:9232)
```
great. Could you open a pull request against the OSGeo/PROJ repository to register this grid in the data/sql/grid_alternatives.sql file ? (mostly a matter to register the fact that the grid will be in the proj-datumgrid-europe package )

dalia2 replied:
Opened the new pull request for grid_alternatives.sql

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+The transformation is designed so that the ISN2016 points should fit within 1-2 cm accuracy, but in the area around the earthquake area in the South and North of Vatnajökull the accuracy is worse.
+
+Recommended to transform from ISN2004 2D geodetic (EPSG:5324) to ISN2016 2D geodetic (EPSG:8086), also this transformation has code in EPSG Dataset (EPSG:9233)
```
same here. To register in PROJ's grid_alternatives.sql file

dalia2 replied:
Opened the new pull request for grid_alternatives.sql

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+Geoid usually fits within 1-2cm, but in some areas it‘s slightly worse 3-5cm. The std for the fitted geoid is 1.7cm based on over 300 gnss levellin points
+
+Grid to convert from ISN2004 ellipsoidal height (EPSG:5323) to national vertical datum ISH2004 (EPSG:8089)
```
not yet registered in EPSG ? You could potentially add a custom entry in the meantime as done in https://github.com/OSGeo/PROJ/pull/1759/files#diff-7d4dfd28cbd3f979bcc5f926e930938fR97

rouault replied:
actually thinking that a better place for custom transformations is in the grid_transformation_custom.sql file at https://github.com/OSGeo/PROJ/blob/master/data/sql/grid_transformation_custom.sql . You will note the use of a 'GravityRelatedHeight to Geographic3D' PROJ-specific method that inverts the confusing convention used by the EPSG's 'Geographic3D to GravityRelatedHeight'. As the .gtx file contains the undulation and all other grid transformations use an additive convention (to go from source to target, one adds the value(s) of the grid), this is far more natural.

dalia2 replied:
I added entries into grid_transformation_custom.sql

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+2D geodetic transformation from Iceland Datum ISN93 to Iceland Datum ISN2016
+
+*Source*: [National Land Survey of Iceland](https://atlas.lmi.is/LmiData/index.php?id=626468364600)
```
For proper display in github, could you make sure that that line (and others starting with '*', except the last one in a series, like "*Credit*" here) are terminated by 2 space character.

dalia2 replied:
Updated the file regarding this

rouault commented on europe/ISL:
```diff
@@ -0,0 +1,312 @@
+
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# LCC Hj�rsey1955 (ellipsoid: International 1924):  The historical systems 2D
```
encoding issue ? I assume we probably want UTF-8

dalia2 replied:
I have changed the ö to o.

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+Deformation model used in transformation between ISN93 reference frame and ISN2016 reference frame
+Horizontal transformation is designed so that the ISN2016 points should fit within 1-2 cm accuracy, but in the area around the earthquake area in the South and North of Vatnajökull the accuracy is worse.
+Vertical transformation 
+Recommended to transform from ISN93 3D geodetic (EPSG:4945) to ISN2016 3D geodetic (EPSG:8085)
```
we could also probably add a custom entry in the PROJ database to register the PROJ pipeline between EPSG:4945 and EPSG:8085. A (really dummy) example of registering a PROJ pipeline string as a 
transformation is in https://github.com/OSGeo/PROJ/blob/master/test/unit/test_factory.cpp#L2074 . To be honest, this hasn't been used yet except in that unit test, so up to you to decide if you want attempt this or not :-)

dalia2 replied:
Yes, we would like to attempt it, here are the lines. Will you put the lines or should I?
EPSG:4945 To EPSG:8085
+proj=pipeline   +step +init=ISL:ISN93_16h   +step +inv +init=ISL:ISN93_16v 
expending will be
 +proj=pipeline   +proj=hgridshift +grids=ISN93_ISN2016.gsb  +step +inv +proj=vgridshift +grids=ISN93_ISN2016.gtx vunits=m 

EPSG:5323 To EPSG:8085
+proj=pipeline   +step +init=ISL:ISN04_16h   +step +inv +init=ISL:ISN04_16v

expending will be
 +proj=pipeline   +proj=hgridshift +grids=ISN2004_ISN2016.gsb  +step +inv +proj=vgridshift +grids=ISN2004_ISN2016.gtx vunits=m

rouault replied:
I'm not sure I'll have the time myself to tackle that. I would strongly suggest using the expended syntax. And perhaps consider this as a follow up action after your initial submission. As I said, I'd consider this as an experimental feature as it has been little tested, so code changes might potentially be also required.

dalia2 replied:
We will skip this approach.

rouault commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,111 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+*License*: [Creative Commons Attribution 4.0](https://creativecommons.org/licenses/by/4.0/)
+*Credit*: (c)2017-2019 Grid created by Guðmundur Valsson.
+
+Used in transformations between ISN2016 and global reference frame realisation ITRF14 with different epochs, 
```
probably nitpicking, but ITRF14 didn't immediately resolve to ITRF2014 in my mind. Everywhere else in PROJ and in EPSG, the name ITRF2014 is used.

kbevers replied:
Yes, ITRF2014 is the official name. I think confusion often arise because we also have IGS14 for GNSS processing, which is almost the same thing as ITRF2014.

dalia2 replied:
Thank you for pointing that out. I have edited the file regarding that.

rouault reviewed (commented):
A few comments. Great to see Iceland join !

rouault commented on travis/expected_europe.lst:
```diff
@@ -47,6 +47,14 @@ ggpf08-Tubuai.gtx
 ggspm06v1.gtx
+icegeoid_2011.gtx
+ISL
+ISN_vel_beta.ct2
```
sorting order raised by https://travis-ci.com/OSGeo/proj-datumgrid/builds/138843342. ISN_vel_beta files should be put after ISN93_ files

dalia2 replied:
I have edited the file.

dalia2 replied:
I uploaded new gtx files: Icegeoid_ISN2004.gtx, Icegeoid_ISN93.gtx and Icegeoid_ISN2016.gtx. I will remove icegeoid_2011.gtx

kbevers commented on europe/ISL:
```diff
@@ -0,0 +1,312 @@
+#------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+#
+#  This file is Work-In-Progress, and should not be used without testing. It is
+#  the official definition and implementation of transformations between different Iceland's systems.
```
This is self-contradictory. These are the official transformations but I shouldn't trust them? @dalia2, you and Gudmundur have done a solid job on these transformations. You should stand by them without any precautions.

dalia2 replied:
I have edited the file regarding that.

kbevers commented:
This is great @dalia2. @rouault have already commented on the few things that need to be fixed before this is ready to merge. I is great that you have added test material inside the `ISL`-file. I haven't yet had the possibility to run them but I will make sure to do so before we merge this.

dalia2 committed:
Update README.EUROPE

dalia2 committed:
Update ISL

dalia2 committed:
Update ISL

dalia2 committed:
Update expected_europe.lst

dalia2 committed:
Update README.EUROPE

dalia2 committed:
Update README.EUROPE

dalia2 committed:
Add files via upload

kbevers commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,145 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+See the `ISL` init-file for examples of use.
+
+* Icegeoid_ISN2004.gtx 
```
As far as I can see this file is not added to the repository. The only file with the "icegeoid" name is `icegeoid_2011.gtx`

kbevers replied:
Fixed in 6c8195d2843860a63b615f8c4398d7dae0fc0a59

kbevers commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,145 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+See the `ISL` init-file for examples of use.
+
+* Icegeoid_ISN93.gtx
```
As far as I can see this file is not added to the repository. The only file with the "icegeoid" name is `icegeoid_2011.gtx`

kbevers replied:
fixed in 6c8195d2843860a63b615f8c4398d7dae0fc0a59

kbevers commented on europe/README.EUROPE:
```diff
@@ -294,3 +294,145 @@ Residuals at 130 further test points average 0.06m, maximum 0.16m.
+
+See the `ISL` init-file for examples of use.
+
+* Icegeoid_ISN2016.gtx
```
As far as I can see this file is not added to the repository. The only file with the "icegeoid" name is `icegeoid_2011.gtx`

kbevers replied:
Fixed in 6c8195d2843860a63b615f8c4398d7dae0fc0a59

dalia2 committed:
Update expected_europe.lst

dalia2 committed:
Update ISL

dalia2 committed:
Delete icegeoid_2011.gtx

dalia2 committed:
Update ISL

rouault commented:
You should rebase on master as there are conflicts. For consistency with current pratice, the text in README.europe should be split into 2 parts. The grids somewhere before the "## Included init-files" section, and the init ISL file after it

dalia2 committed:
Merge branch 'master' into add-grids-for-Iceland

kbevers commented on europe/ISL:
```diff
@@ -0,0 +1,388 @@
+#  
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+#
+#   <metadata> +version=1.0.0 +origin=LMI +lastupdate=2019-11-30
```
This won't get picked up by PROJ since it is commented out

kbevers commented on europe/ISL:
```diff
@@ -0,0 +1,388 @@
+#  Deformation grids 
+#--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+<ISN93_16h> +proj=hgridshift +grids=ISN93_ISN2016.gsb
+<ISN93_16v> +proj=vgridshift +grids=ISN93_ISN2016.gtx vunits=m 
```
Remove use of PROJ.4 modifier that doesn't effect the output
```suggestion
<ISN93_16v> +proj=vgridshift +grids=ISN93_ISN2016.gtx
```

kbevers commented on europe/ISL:
```diff
@@ -0,0 +1,388 @@
+#----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# LCC ISN2016 (epoch 2016.5, IGS2014, GRS80):  The primary system  
+#----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+<ISN2016_LAM>  +proj=lcc  +lat_1=64.25 +lat_2=65.75 +lat_0=65 +lon_0=-19 +x_0=2700000 +y_0=300000  +ellps=GRS80   +units=m +no_defs
```
Remove use of PROJ.4 modifiers that doesn't effect the output
```suggestion
<ISN2016_LAM>  +proj=lcc  +lat_1=64.25 +lat_2=65.75 +lat_0=65 +lon_0=-19 +x_0=2700000 +y_0=300000  +ellps=GRS80
```

kbevers commented on europe/ISL:
```diff
@@ -0,0 +1,388 @@
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# LCC Hjorsey1955 (ellipsoid: International 1924):  The historical systems 2D
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+<HJORSEY_LAM> +proj=lcc   +lat_1=65  +lat_0=65 +lon_0=-18 +x_0=-500000 +y_0=500000 +ellps=intl +units=m   +no_defs +axis=wnu
```
Remove use of PROJ.4 modifier that doesn't effect the output
```suggestion
<HJORSEY_LAM> +proj=pipeline 
				+step +proj=lcc   +lat_1=65  +lat_0=65 +lon_0=-18 +x_0=-500000 +y_0=500000 +ellps=intl
				+step +proj=axisswap +order=-1,2,3,4
```

kbevers reviewed (changes requested):
I finally had some time to properly go through the init file. There's a few things that I would suggest be changed. 

Use of deprecated PROJ.4 +modifiers should be avoided. I see heave use of `+units`, `+vunits`, `+no_defs` and `+axis`. Those should all be either removed or replaced by a an extra pipeline-step. `+units=m`, `+vunits=m` and `+no_defs` can generally be removed without consequences. If other units than meters are needed a `+proj=unitconvert` pipeline step should be inserted. `+axis=wnu` should be replace by a `+proj=axisswap +order=-1,2,3,4` pipeline-step. I've added a few examples in in-lined comments to show how to do it.

I notice that all of the pipelines in the file are written on one long line. You can improve readability by putting each step on a line by itself. There's an example in the in-lined comments.

Furthermore, the gie tests in-lined in `ISL` does not pass. Putting in the tests is great use of the format but it is not helpful unless the tests can pass. I get the following output from my gie run:

```
gie europe\ISL
-------------------------------------------------------------------------------
Reading file 'europe\ISL'
     -----
     FAILURE in ISL(166):
     expected: 357557.5871 407700.8896
     got:      -38072.224780227290   -8122266.058058928698
     deviation:  999999999.999000 mm,  expected:  50000.000000 mm
     -----
     FAILURE in ISL(173):
     expected: 357557.5871 407700.8896
     got:      -38072.224780226708   -8122266.058058928698
     deviation:  999999999.999000 mm,  expected:  50000.000000 mm
     -----
     FAILURE in ISL(186):
     expected: 354399.1771 409035.1420
     got:      -49792.425436145626   -8121716.680567346513
     deviation:  999999999.999000 mm,  expected:  15000.000000 mm
     -----
     FAILURE in ISL(217):
     expected: 1745293.0413 470683.5387 75.6532
     got:      inf   inf   inf
     deviation:  999999999.999000 mm,  expected:  50.000000 mm
proj_create: Error -1: no arguments in initialization list
-------------------------------------------------------------------------------
init=ISL:ISN2016_ISH2004_v
-------------------------------------------------------------------------------
Invalid operation definition in line no. 350:
       no arguments in initialization list (errno=no_args/-1)
     -----
     FAILURE in ISL(370):
     expected: 4740331.7212 376698.7742 134.2005
     got:      4740331.696816119365   376698.830172262969   134.208136844
     deviation:  61.528767 mm,  expected:  30.000000 mm
     -----
     FAILURE in ISL(387):
     expected: 2740331.7315 376698.7505 134.1972
     got:      2740331.686488890555   376698.853889305959   134.211391162
     deviation:  113.651827 mm,  expected:  30.000000 mm
-------------------------------------------------------------------------------
total: 15 tests succeeded,  0 tests skipped,  7 tests FAILED!
-------------------------------------------------------------------------------
```

dalia2 committed:
Update ISL

dalia2 commented:
Thank you for your reviews, and sorry for mistakes. I updated ISL file using your suggestions.
Regarding the failures of gie. I think something is wrong with gie itself, it calculates wrongly. Or am I doing something wrong regarding gie? Should I delete these three examples with gie?

 FAILURE in ISL(166), i put  a line in PROJ:
 echo  -21.930875   64.1421888889 0 0  | cct    +proj=pipeline  +step +proj=cart   +ellps=danish  +step +init=c:\PROJ\ISL:REYK19_ISN93   +step +proj=cart  +inv +ellps=GRS80  +step +init=c:\PROJ\ISL:ISN93_LAM
**output : 357557.5871  407700.8896**  

FAILURE in ISL(173):
echo -34.50875   64.1421888889  0 0   | cct  +proj=pipeline  +step +inv +proj=longlat  +pm=copenhagen  +step +proj=cart   +ellps=danish  +step  +init=c:\PROJ\ISL:REYK19_ISN93 +step +proj=cart  +inv +ellps=GRS80   +step +init=c:\PROJ\ISL:ISN93_LAM 
**output: 357557.5871  407700.8896**

FAILURE in ISL(186):
 echo -21.9933555556   64.1533913888889  0 0   | cct +proj=pipeline +step +proj=cart +ellps=intl +step  +init=c:\PROJ\ISL:HJ_ISN93  +step +proj=cart +inv +ellps=GRS80 +step    +init=c:\PROJ\ISL:ISN93_LAM
**output: 354399.1771  409035.1420** 

 FAILURE in ISL(217):
I fixed it

FAILURE with init=ISL:ISN2016_ISH2004_v
I fixed it. 

FAILURE in ISL(370):
I fixed it

FAILURE in ISL(387):
I fixed it

In some cases I am using this expression (Example: echo  -34.50875   64.1421888889 0 0  | cct  init=ISL:RJ_W), is this syntax correct?:

dalia2 committed:
Update ISL

dalia2 committed:
Update ISL

kbevers reviewed (commented):
>  I think something is wrong with gie itself, it calculates wrongly. Or am I doing something wrong regarding gie? 

gie is not calculating incorrectly but it may not be easy to see why that is. The first failure is a test if `ISL:RJ_G`. If I run the test manually, you'll see that the height is changed:

```
echo -21.930875   64.1421888889 0 0 | cct +init=ISL:RJ_G
  357557.5871    407700.8896       65.7290        0.0000
```
That is because you are doing a conversion to and from cartesian coordinates which makes your transformation a 3D transformation. You can get by the problem either by modifying the gie test:

```
<gie>
operation   init=ISL:RJ_G
tolerance   50 m
accept      -21.930875   64.1421888889   0.0
expect      357557.5871   407700.8896    65.7290
</gie>
```

or by adjusting the pipeline so that it works better with 2D coordinates:

```
<RJ_G> proj=pipeline
    step proj=push +v_3
    step proj=cart   ellps=danish
    step init=ISL:REYK19_ISN93
    step proj=cart  inv ellps=GRS80
    step proj=pop +v_3
    step init=ISL:ISN93_LAM
```

Here I am forcing PROJ to ignore the height and set it to whatever it was in the input coordinate. With this trick you still have to modify the test though since it is still a 3D transformation:

```
<gie>
operation   init=ISL:RJ_G
tolerance   50 m
accept      -21.930875   64.1421888889 0
expect      357557.5871   407700.8896  0
</gie>
```

What the best course of action is is for you to decide. I don't know if the height transformation is important or not. The above is the cause for all three failures.

> Should I delete these three examples with gie?

No, just make them work as per the above. It is very valuable to have the tests there as you can use it to check the PROJ works as expected when new versions come out.

dalia2 committed:
Update ISL

kbevers commented:
With the latest commit I think we have sorted all that needed to be fixed. Good job, @dalia2. Are you happy with this as it is now?

dalia2 commented:
Yes, i am happy with this as it is. Thank you for your help.

rouault commented:
@dalia2 May I ask a complementary work ? In provision for a future mass conversion to GeoTIFF, could you update https://github.com/OSGeo/proj-datumgrid/blob/master/filelist.csv to add entries for the new files ? At least for the classic geoid models. I'm not yet sure about how we will deal with the .ct2 + .gtx files used for the velocity models.

rouault commented:
and  regarding filelist.csv, the NTv2 grids as well
